### ISSUE TITLE
Generated R script is not working

### ISSUE THREAD

**Issue by aalfons** [VLDrenth/robmedShiny#25]
This is an example of an R script that I generated with the GUI:

```
load('/Library/Frameworks/R.framework/Versions/4.2/Resources/library/robmedExtra/shiny-examples/GUI/BSG2014.Rdata')
control_var <- reg_control(efficiency =  0.85 , max_iterations =  10000 )

formula_model_1 <- TeamCommitment ~ m(TaskConflict, .model = "parallel") + ValueDiversity
bootstrap_test_1 <- test_mediationrormula_model1, data = BSG2014, robust = TRUE,level = 0.95, control = control_var)



formula_model_2<- TeamCommitment ~ m(TaskConflict, .model = "parallel") + ValueDiversity
bootstrap_test_2<- test_mediation(formula_model2, data = BSG2014, robust = FALSE,level = 0.95)
```

It is not working for several reasons:

1. The script doesn't include a line to load package `robmed`, hence the `reg_control()` command on line 2 produces an error.
2. When using an existing data set from the global environment, the shiny app saves the data in the folder where the shiny app lives, which is a subfolder where the package is installed.
    - The user may not have write permissions to that folder.
    - In this example, I used the `BSG2014` example data from the `robmed` package. The shiny app saved a file `BSG2014.RData`, but the object that is stored in the `RData` file is called `df`, not `BSG2014`. But the commands for `test_mediation()` use the argument `data = BSG2014`, so that command will fail.
3. The commands for `test_mediation()` are not written correctly to the R script. This is very obvious for ROBMED, but also the command for the OLS bootstrap uses `formula_model2` although the defined formula object is called `formula_model_2`.

**Comment by aalfons** [comment]
Regarding 3.: Currently, creating the output for the shiny app and creating the commands that are put in the R script is done in two different places. This is a very dangerous practice! Not only is it prone to introduce bugs in the R script, it also makes maintenance of the shiny app more difficult. If the command to produce the output changes for some reason in a future version, we need to remember to also change the command that is put in the R script.

This can be avoided. We can create the command as an expression, for example:
```
command <- expression(
  boot <- test_mediation(BSG2014, 
                         x = "ValueDiversity", 
                         y = "TeamCommitment",
                         m = "TaskConflict")
)
```
Then we can use `eval(command)` to run the command for creating the output, and we can use `as.character(command)` to convert the command to a text string that can later be exported into the R script.

@VLDrenth, this is rather advanced R programming, and I should have mentioned something about this when we first discussed generating the R script. So this one is my fault, sorry! In the future, if you're unsure about how to do something in R or if you're unsure if your solution is a good one, please just ask. You don't have to figure everything out on your own.

**Comment by VLDrenth** [comment]
@aalfons Thank you letting me know about this technique! One question about the use of that in shiny: functions generally depend on some input or a reactive expression in the app, e.g. for getting the formula. This would not work in the script as it is outside the reactive environment, but when storing a command as expression() it saves the command with formula() , rather than the formula type value that we would like it to. What would you suggest to resolve that issue?

**Comment by aalfons** [comment]
It's actually better not to use expression, but to build up the function call with `call()`. For example, if we have a list of inputs
```
input <- list(df_name = "BSG2014", x = "ValueDiversity", y = "TeamCommitment", m = "TaskConflict")
```
we can build the function call with
```
command <- call("test_mediation", as.name(input$df_name), x = input$x, y = input$y, m = input$m)
if (!is.null(input$covariates)) command$covariates <- input$covariates
```
The result looks like this
```
test_mediation(BSG2014, x = "ValueDiversity", y = "TeamCommitment", 
    m = "TaskConflict")
```
Note that the name of the function needs to be given as a text string. More importantly, the text string with the name of the data frame needs to be converted with `as.name()` so that it shows up without quotes in the function call. R stores function calls as objects of class `"call"`. Those are essentially lists, which allows us to modify the function call in the same way we do lists, which is very convenient to keep the function call clean if the user did not set a certain input (in the example, the `covariates` argument).

We can evaluate the function call with `eval(command)` to generate the output in the shiny app, and we can use `as.character(as.expression(command))` to convert it to a text string to be exported to the R script.

**Comment by aalfons** [comment]
I suggest not to use `robmed`'s formula interface to build up the function call, since we can use the inputs more directly if we use the default method of `test_mediation()` with arguments `x`, `y`, `m`, and so on. It also makes the exported R script a bit cleaner (currently, separate formula objects are generated for different calls to `test_mediation()` even if the formula is the same).

**Comment by VLDrenth** [comment]
@aalfons I think in the latest commit everything should be correctly implemented regarding the script with using call. Could you take a look to verify if it looks alright? Also a question regarding the script: do we only include the calls to test_mediation or also things like generating scripts and/or tables?

**Comment by aalfons** [comment]
Thanks, the commands now work, except that there is no line for loading the `.RData` file. I only looked at the script itself, not yet at the underlying code.

But a few things are missing from the script:
1. We need to assign the output to an object (e.g., `robust_boot` and `ols_boot`) and then add a line that creates the summaries (e.g., `summary(robust_boot)`). We'll need to look into how to do this in a safe manner based on the function call we created. Perhaps we can modify it further after converting to an expression.
2. There should be an empty line after different steps. But empty lines should always be created before adding new code to the script, not after adding code (such that the generated script doesn't have multiple empty lines at the end).
3. It should be `library("robmed")` not `library('robmed')`. We should be consistent about which quotes we are using for text strings - single or double. My preference is double quotes.
4. We should also add comments in the script, e.g., `# load package`, `# apply the robust bootstrap test ROBMED`, `# apply the OLS bootstrap test`.

We should also talk about what to do with the data set. 
- When someone loads an `.RData` file, we can easily add the corresponding line to the script. 
- When someone uses an existing data frame from the global environment, I think we should just add a comment in the script that the code to load the data needs to be added by the user to have a reproducible script.

About your question, the end goal is indeed to add also the code for generating the tables and saving the figure. That will make the output (which can be used in a paper) fully reproducible. We don't include the code to download the script itself in the script, as that would be circular.
